### PR TITLE
[Calyx] Verify CombGroupOp is combinational and certain primitive ports are driven together.

### LINKED ISSUE #1712: [Calyx] Verify certains ports are written or driven together.
A few examples:
- [x] A MemoryOp should only be read/written to if its address ports are driven.
    #1764
- [x] A RegisterOp should only be written to if the `write_en` port is driven.
    #1764
- [x] A binary operation should only be read if both the `lhs` and `rhs` have been driven (similar for unary operation). For this one, it could *technically* be driven in another group, but that would require us to walk the ancestor groups in the control each time we verify. A simpler, more restrictive verification is just ensure that `left` and `right` are driven within the same group.
    #1764

Calyx does this by means of an attribute in [Calyx 623](https://github.com/cucapra/calyx/pull/623). This could be done by means of attribute, but I imagine just having this as an addition to `verifyGroupOp` is fine as well.

### PR DESCRIPTION
**Pull request by cgyurgyik**
This PR focuses on (2) objectives:
1. Verifying that combinational groups are indeed combinational. Combinational components are not supported yet, so component instances are always assumed to be combinational for now. 
2. We also verify that certain ports are written or read together. For example, it does not make sense to write to a register without `write_en` being high. To aid in this, I added helper functions to access certain ports on primitive operations.

While combinational operations may be driven in different groups (i.e. `.left` in one group, `.right` in another), I've actually never seen this in practice. For now, I think its fine to keep this, and remove it down the road if it is too restrictive.

Closes #1712.

### CODE REVIEW

File: lib/Dialect/Calyx/CalyxOps.cpp [hunk 1]
```diff
@@ -186,6 +219,105 @@ LogicalResult calyx::verifyCell(Operation *op) {
...
+  if (destDefiningOp == nullptr)
+    return success();
+  auto destCell = dyn_cast<CellInterface>(destDefiningOp);
+  if (destCell == nullptr)
+    return success();
+
+  LogicalResult verifyWrites =
+      TypeSwitch<Operation *, LogicalResult>(destCell)
+          .Case<RegisterOp>([&](auto op) {
+            // We only want to verify this is written to if
```
Comment (mikeurbach) [inline lib/Dialect/Calyx/CalyxOps.cpp:236]: This sentence seems unfinished.

**Review by mikeurbach — APPROVED** [review]
Looks great! I love how much semantics we're starting to capture for Calyx ops. The more guardrails we put in now, the happier we will be when we target Calyx later.

File: lib/Dialect/Calyx/CalyxOps.cpp (hunk 1 above)
Comment (cgyurgyik) [reply lib/Dialect/Calyx/CalyxOps.cpp]: Whoops, unintended mad lib.

File: lib/Dialect/Calyx/CalyxOps.cpp [hunk 2]
```diff
@@ -594,6 +727,40 @@ static LogicalResult verifyWiresOp(WiresOp wires) {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// CombGroupOp
+//===----------------------------------------------------------------------===//
+
+/// Verifies a combinational group may contain only combinational primitives.
+// TODO(www.github.com/llvm/circt/issues/1739): Add Combinational trait.
```
Comment (rachitnigam) [inline lib/Dialect/Calyx/CalyxOps.cpp:735]: I don’t know if this is correct. For example, it is perfectly okay to read a value from a register in a combinational group. Similarly, memories may be used in combinational groups as well if they’re being read.

Comment (cgyurgyik) [reply lib/Dialect/Calyx/CalyxOps.cpp:735]: Can you provide some more insight? I'm happy to remove this, but there is some misunderstanding on my part. In Calyx, I know we always consider register writes to be one cycle for simplicity, given that it is a non-combinational circuit. For memory, we consider memory writes to be a single cycle as well. I imagine a memory load would *at least* take a single cycle. So to read from a memory and then use it in a combinational group, the control might look like:

```mlir
calyx.control {
  calyx.seq {
    calyx.enable @LoadFromMemory {static=N} // Read the intended value.
    calyx.if @CombGroup { ... }             // Use the value combinationally.
  }
}
```

In other words, why can we get away with using non-combinational modules in combinational groups?

Comment (rachitnigam) [reply lib/Dialect/Calyx/CalyxOps.cpp:735]: The problem with this verification check is that while combinational primitives (`comb primitive`) only have combinational circuit paths in them, a non-combinational primitive can have *both* combinational and sequential paths in it.

For example, the implementation of the memory primitive in Calyx takes one cycle to write a value into memory but can combinationally read the memory. So the following combinational groups are all fine (and indeed show up when generating code from Dahlia):

```
comb group read_reg {
  eq.left = reg.out;
  eq.right = 32'd10;
}

comb group read_mem {
  mem.addr0 = 4'd2;
  eq.left = mem.read_data;
  eq.right = 32'd10;
}
```

It is worth noting that different memory primitives may implement reads sequentially (i.e. take one or more cycles to return the value of a read). 

TLDR: Non-combinational primitives may be used in combinational groups as long as the ports they're been read from are on a combinational circuit path.

Comment (cgyurgyik) [reply lib/Dialect/Calyx/CalyxOps.cpp:735]: Ok so just to clarify, you're implying that current memory primitives in Calyx (RegisterOp, MemoryOp) conduct reads combinationally.

Comment (rachitnigam) [reply lib/Dialect/Calyx/CalyxOps.cpp:735]: Registers always provide reads combinationally. Memories may provide reads combinationally.

Comment (cgyurgyik) [reply lib/Dialect/Calyx/CalyxOps.cpp]: Fixed this. Thanks for the context.

**Comment by cgyurgyik** [comment]
Going to push this @mortbopet since I have time to refactor the other open PR. As always, I will address any comments you leave post-submit.
